src/cls/rbd/cls_rbd_client.cc

```diff
@@ -13,36 +13,43 @@
 namespace librbd {
   namespace cls_client {
 
-    void get_immutable_metadata_start(librados::ObjectReadOperation *op) {
-      bufferlist bl, empty_bl;
+    void get_initial_metadata_start(librados::ObjectReadOperation *op) {
+      bufferlist bl, empty_bl, features_bl;
       snapid_t snap = CEPH_NOSNAP;
       ::encode(snap, bl);
       op->exec("rbd", "get_size", bl);
       op->exec("rbd", "get_object_prefix", empty_bl);
+
+      ::encode(snap, features_bl);
+      ::encode(true, features_bl);
+      op->exec("rbd", "get_features", bl);
     }
 
-    int get_immutable_metadata_finish(bufferlist::iterator *it,
+    int get_initial_metadata_finish(bufferlist::iterator *it,
                                       std::string *object_prefix,
-                                      uint8_t *order) {
+                                      uint8_t *order,
+                                      uint64_t *features) {
       try {
 	uint64_t size;
 	// get_size
 	::decode(*order, *it);
 	::decode(size, *it);
 	// get_object_prefix
 	::decode(*object_prefix, *it);
+	// get_features
+	::decode(*features, *it);
```

Comment:
Note, `get_features` returns also `incompatible_features`. It is not necessary to decode this field here, because it is the last item. On the other hand, if someone later adds yet another `get_xyz` after `get_features` he might forget (not know) to decode `incompatible_features`, so probably it makes sense to decode it (or left a comment). I don't have a strong opinion here though.

Reply:
I'd vote for decoding it into a throwaway local variable for clarity

Reply:
Still looks like this comment is open

Reply:
@dillaman decode `features` for test_features.

Reply:
@gmayyyha We meant that you need to add something like this:
```
uint64_t incompatible_features;
::decode(incompatible_features, *it);
```
See "get_size" case above where we decode size in a throwaway local variable.

Reply:
@trociny Ah...  Sorry, I misunderstand. fixed.